fuse/dl/cross_validation/pipeline.py

```diff
@@ -0,0 +1,303 @@
+    train_func: Callable,
+    infer_func: Callable,
+    eval_func: Callable,
+    dataset_params: Dict = None,
+    train_params: Dict = None,
+    infer_params: Dict = None,
+    eval_params: Dict = None,
+    paths: Dict = None,
+    deterministic_mode: bool = True,
+    sample_ids_per_fold: Sequence = None,
+) -> None:
+    """
+    ML pipeline - run a full ML experiment pipeline which consists of training on multiple
+    cross validation folds, validation inference and evaluation, inference and evaluation
+    of each fold's model on a held out test set, ensembling the models trained on different folds,
+    and evaluation of the ensembled model.
+    The whole process can also be repeated multiple times, to evaluate variability due to different
+    random seed.
+    Multiple cross validation folds run simultaneously on available GPU resources.
+
+    :param num_folds: Number of cross validation splits/folds.
+    :param num_folds_used: Number of folds/splits to use.
+        For example, for training a single model with 80% of the samples used
+        for training and 20% for validation, set `num_folds=5` and `num_folds_used=1`.
+        For running a 5-fold cross-validation scenario, set `num_folds=5` and `num_folds_used=5`.
+    :param num_gpus_total: Number of GPUs to use in total for executing the pipeline.
+    :param num_gpus_per_split: Number of GPUs to use for a single model training/inference.
+    :param num_repetitions: Number of repetitions of the procedure with different random
+        seeds. Note that this does not change the random decision on cross validation
+        fold sample ids.
+    :param dataset_func: Callable to a custom function that implements a dataset
+        creation. Its input is a path to cache directory and it should return a
+        train and test dataset.
+    :param train_func: Callable to a custom function that executes model training.
+    :param infer_func: Callable to a custom inference function.
+    :param eval_func: Callable to a custom evaluation function.
+    :param dataset_params: Dictionary that can contain any number of additional
+        custom parameters for dataset_func.
+    :param train_params: Dictionary that can contain any number of additional
+        custom parameters for train_func.
+    :param infer_params: Dictionary that can contain any number of additional
+        custom parameters for infer_func.
+    :param eval_params: Dictionary that can contain any number of additional
+        custom parameters for eval_func.
+    :param deterministic_mode: Whether to use PyTotch deterministic mode.
+    :param sample_ids_per_fold: May contain a sequence of array pairs denoting sample ids
+        for pre-defined train/validation splits. If this parameter is set to `None`,
+        the splits are decided at random.
+    """
+
+    if deterministic_mode:
+        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"  # required for pytorch deterministic mode
```

Comment:
Maybe we should add it to Seed.set_seed function?

Reply:
done